### ISSUE TITLE
Issues with custom dataset.

### ISSUE THREAD

**Issue by bhosalems** [voxelmorph/voxelmorph#468]
I have been trying to do pairwise registration on the CHAOS dataset. https://chaos.grand-challenge.org/Data/

First I converted the volumes from .dcm to .nii. I converted the segmentation labels from .png to .nii. and chose all the permutations of volumes as fixed and moving images.

I am not sure why my registered/warped moving image has a very different intensity distribution than the original image. Sometimes with erroneous hollow patches. Please check the outputs below, the first image is the fixed image, the second image is the moving image and the third image is the registered moving image. This is causing very low dice scores (<10%) although similarity scores are in the range of 60-70%.
![image](https://user-images.githubusercontent.com/10846405/187094929-8cb6440f-eac3-4c13-8310-cf214547c3c3.png)
![image](https://user-images.githubusercontent.com/10846405/187094941-952e7cb1-f125-46e5-836f-60f19d47c175.png)
![image](https://user-images.githubusercontent.com/10846405/187094956-559d5bfd-52a0-4eda-9d13-7fb45ffa0c6f.png)

I have trained the same model on MSD http://medicaldecathlon.com/ and it works just fine. What am I doing wrong or what parts of code I should look into (e.g. image processing, spatial transformers, unet etc)?

Thanks

**Comment by adalca** [comment]
I wonder if this is a visualizlation/clipping issue rather than a deformation issue. Can you visualize the deformation field (e.g. in 3 different channels)? How are you normalizing the data?

**Comment by bhosalems** [comment]
Thanks for the response.

Below is the code for normalization -
```
mean = np.mean(data)
std = np.std(data, ddof=1)
maxp = mean + 6*std
minp = mean - 6*std
y = np.clip(data, minp, maxp)
z = (y-y.min())/y.max()
data = z
```
Let me try to visualize the deformation field.

**Comment by adalca** [comment]
are you sure the data is float? 

also this might be minor but im not sure if that normalization makes sense since you're using the max of y and not (y-y.min())...
But this wouldn't lead to the crazy image at the end I don't think

**Comment by bhosalems** [comment]
Yes, the data is float.

I don't understand your comment about normalization. I tried to look into the repo for reference but I couldn't locate anything related to normalization/data loading.

Also, would you mind pointing to the file which visualizes the flow in 3 channels?

Thanks.

**Comment by bhosalems** [comment]
Apologies for the delayed reply, so I was able to visualize the flow -
![flow](https://user-images.githubusercontent.com/10846405/188032693-2cb1b041-32f9-4f28-916d-0b53d04ea8f8.png)
Corresponding fixed and warped images are,
![image](https://user-images.githubusercontent.com/10846405/188032805-70f8e9fd-b89b-44eb-8985-ef4d86a34713.png)
![image](https://user-images.githubusercontent.com/10846405/188032764-a3c9b7f2-ce9b-4545-b296-449e59c79f71.png)
It does look like there's something going on with the deformation field.
FYI, I have used code from https://github.com/junyuchen245/TransMorph_Transformer_for_Medical_Image_Registration/issues/14 to visualize the flow field.

**Comment by adalca** [comment]
re:normaliation -- you're dividing by the max of the wrong thing, I think. Say y.min is 10000 and y.max is 10001 (i made that up, but as an example). then y-y.min is around [0, 1], but then you're dividing it by 10001, which will be a small number.

re:deformation visualization: the field actually looks quite smooth outside of the anatomy, and yet when you warp the anatomy you get this crazy noisy artifact. I'm guessing this is an issue with the data type or visualization of the images, no?

I would recommend visualizing the field a bit differently -- first draw a black-and-white grid, and then warp it with the deformation field. That will show the dielf much better, in my opinion. bw grid here:
https://github.com/adalca/pystrum/blob/0e7a47e5cc62725dfadc728351b89162defca696/pystrum/pynd/ndutils.py#L111

**Comment by bhosalems** [comment]
Thanks, below is one of the flows warped with be grid.
![e17idx258_m37f36grid_flow](https://user-images.githubusercontent.com/10846405/190156885-c103a2b4-ab51-4258-b810-437252fcbdd1.png)

**Comment by adalca** [comment]
and what do you think about it?

**Comment by bhosalems** [comment]
that there is some flow even outside volume and I am unaware of what could be causing this. One of the things that is showing improvement in the train dice scores (at least from the initial results) is increasing the size of the window for calculating the similarity/NCC loss. Maybe because we are expecting to have large deformations or the size of the organs is large given this is an abdominal MRI dataset (vs. say brain MRI) increasing the window size is helpful. I haven't looked at the latest warped deformation field yet (as it is not trained enough yet), but likely should be better than the previous ones?

**Comment by adalca** [comment]
the flow outside the volume might just be an interpolation artifact. To me the warps look like the volumes ar enot affinely aligned, which is why the overall shift. Then, inside, the deformations don't look too crazy.

what are the 8 panels?

**Comment by bhosalems** [comment]
Sorry, I forgot to mention the details of the image. Those are slices along the depth. Starting at 8th and spaced by 8 slices from the top left corner to the bottom right corner.

**Comment by bhosalems** [comment]
Yes, it does have some affine misalignment issue. Let me open new issue for the help.
